keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1503 @@
+- Fully qualified domain name
+- Integer (as opposed to any floating point)
+
+> Until CRD Validation Ratcheting is stable, these formats will be translated into an inline schema containing either the regex pattern or the CEL expression for the format.
+
+### IDL tags
+
+IDL tags may be used directly on type declarations and indirectly on field and
+type aliases. For example:
+
+```go
+type Widget struct {
+  // +validationRule="self.matches('[a-z][1-9]+')"
+  Component PartId `json...`
+}
+
+// +maxLength=20
+type PartId = Identifier
+
+// +format=dns1123label
+type Identifier string
+
+type Contraption struct {
+  Component Identifier `json...`
+}
+```
+
+In the above example, the `widget.component` field is validated against all three
+IDL tags. But `contraption.component` is only validated against `+format=dns1123label`.
+
+Shared types present a challenge. For example, different Kubernetes resources
+have different validation rules for `metadata.name` and `metadata.generateName`.
+But all resources share the `ObjectMeta` type.
+
+We can support these uses cases with CEL validation rules:
+
+```go
+type ExampleSpec struct {
+
+  // +validationRule="!has(self.name) || self.name.isFormat('dns1123subdomain')",field="name"
+  // +validationRule="!has(self.generateName) || self.generateName.replace('-$', 'a').isFormat('dns1123subdomain')",field="generateName"
+  metav1.ObjectMeta `json...`
+}
+```
+
+Because the above use case is so common, we plan to offer a special tag to make
+declaring the above rules convenient:
+
+```go
+// +nameFormat='dns1123subdomain'
+metav1.ObjectMeta `json...`
+```
+
+Another example: Pod `container` and `initContainer` fields share the same
+`Container` type but have different validation rules. The rules that are the
+same for both can be declared on `Container` but any rules that are different
+can be declared on the `container` and `initContainer` fields using CEL. E.g.:
+
+```go
+// +validationRule="!has(self.RestartPolicy)"
+Containers []Container `json...`
+```
+
+## Analysis of existing validation rules
+
+At the time of writing this document, there are 1181 validation rules written in about
+15k lines of go code in [kubernetes/kubernetes/pkg/apis](https://github.com/kubernetes/kubernetes/commit/0c62b122c02bff9131b6db960042150a3638d3f3).
+
+Roughly 30% of validation rules may require CEL expressions. 15% of these are
+forbidden rules which primarily check which fields are allowed when a union
+discriminator is set. 10% are object name validations, and the remaining 10% are
+cross field validation checks, mainly mutual exclusion and some "transition
+rules" (e.g. immutability checks).
+
+The remaining 70% of validation rules can be represented using JSON Schema value
+validations. `optional`, `format` and `enum` will the the most frequently used.
+
+![Validation types](validation-types.svg)
+
+![Object metadata name validations](metadata-name-types.svg)
+
+![Requires CEL](requires-cel.svg)
```

Comment:
I'm not sure what "requires CEL" means?

Reply:
changed to include note that the expressions are complicated and thus require CEL